data-generation/generate-instructions.py

```diff
@@ -0,0 +1,204 @@
+from random import randint, random
+
+
+"""
+Classes
+"""
+
+
+class NoActionException(Exception):
+    pass
+
+
+class Block:
+    def __init__(self, side1_letter,
+                 side1_color,
+                 side2_letter,
+                 side2_color,
+                 pos):
+        self.side1_letter = side1_letter
+        self.side1_color = side1_color
+        self.side2_letter = side2_letter
+        self.side2_color = side2_color
+        self.position = pos
+
+    def __eq__(self, other):
+        return self.side1_letter == other.side1_letter and \
+            self.side2_letter == other.side2_letter    and \
+            self.side1_color == other.side1_color      and \
+            self.side2_color == other.side2_color      or  \
+                                                           \
+            self.side1_letter == other.side2_letter    and \
+            self.side2_letter == other.side1_letter    and \
+            self.side1_color == other.side2_color      and \
+            self.side2_color == other.side1_color
+
+    def __ne__(self, other):
+        return not self == other
+
+    def shift_to(self, position):
+        return Block(self.side1_letter,
+                     self.side1_color,
+                     self.side2_letter,
+                     self.side2_color,
+                     position)
+
+    def flip(self):
+        return Block(self.side2_letter,
+                     self.side2_color,
+                     self.side1_letter,
+                     self.side1_color,
+                     self.position)
+
+    def __str__(self):
+        return '{side1: (%s, %s), side2: (%s, %s), pos: %s}' % \
+            (self.side1_letter, self.side1_color,
+             self.side2_letter, self.side2_color,
+             self.position)
+
+
+class Instruction:
+    def __init__(self, phrase, point):
+        self.phrase = phrase
+        self.point = point
+
+    def __str__(self):
+        return self.phrase
+
+
+class Configuration:
+    def __init__(self, blocks, final_blocks=[]):
+        self.current_blocks = blocks
+        self.final_blocks = final_blocks
+
+    def is_complete(self):
+        return self.current_blocks == []
+
+    def get_instruction(self, moved_block):
+        point = moved_block.position
+        phrase = 'move the %s %s block here: %s' \
+                 % (moved_block.side1_color,
+                    moved_block.side1_letter,
+                    point)
+        return Instruction(phrase, point)
+
+    def get_action(self, goal):
+        if self.is_complete():
+            raise NoActionException('Board is already complete')
+
+        block_to_move = rand_element(self.current_blocks)
+        moved_block = goal.final_blocks[goal.final_blocks.index(block_to_move)]
```

Comment:
What exactly is the point of this line? It looks like we're retrieving the identical block in the `goal.final_blocks`, but do we actually need the reference for things like `remove` to work? I could totally just not know Python, too...

Reply:
Yeah, so this is pretty confusing. This is really an abuse of how I define equality, and I think that I should reconsider how this is done (when it isn't past midnight). The two blocks aren't identical, the block in `final_blocks` is in the correct, final position. I can look up that block's index using the unmoved block because block equality does not take position in to account.